## ひろしまQuest2020#stayhome【コース予測部門】

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import lightgbm as lgb
import gc
%matplotlib inline
pd.set_option('display.max_Columns', 100)

In [2]:
ALL_MERGE = 'intermediate/all_merge_16.f'
SUBMIT = 'submit/course_16.csv'

In [3]:
all_pitch = pd.read_feather(ALL_MERGE)
all_pitch.shape

(778767, 181)

In [4]:
print(all_pitch.shape)
column_cnt = len(all_pitch.columns)
all_pitch.head()

(778767, 181)


No  ball  course  No_in_game  pitch_cnt_in_game  inning  bat_cnt_in_inning  \
0   0   0.0     0.0           1                  1       1                  1   
1   1   0.0     8.0           2                  2       1                  1   
2   2   0.0     5.0           3                  3       1                  1   
3   3   0.0    12.0           4                  4       1                  2   
4   4   0.0     8.0           5                  5       1                  2   

   pitch_cnt_in_bat  pitch_order  player_cnt_in_game  pitcher_cnt_in_game  \
0                 1            1                   1                    1   
1                 2            1                   1                    2   
2                 3            1                   1                    3   
3                 1            1                   2                    4   
4                 2            1                   2                    5   

   pitcher_cnt_in_inning  bat_order  bat_cnt_in_game  home_point  away_point  \
0                      1          1                1           0           0   
1                      2          1                1           0           0   
2                      3          1                1           0           0   
3                      4          2                1           0           0   
4                      5          2                1           0           0   

   out_cnt  ball_cnt  strike_cnt  pitch_LR  bat_LR  bc_straight  bc_curve  \
0        0         0           0       0.0     1.0     0.464734  0.110359   
1        0         0           1       0.0     1.0     0.442351  0.093302   
2        0         0           2       0.0     1.0     0.504212  0.069963   
3        1         0           0       0.0     0.0     0.475961  0.101792   
4        1         0           1       0.0     0.0     0.463496  0.099763   

   bc_slider  bc_shoot   bc_fork  bc_changeup  bc_sinker  bc_cutball  \
0   0.151086  0.097383  0.067492     0.043210   0.012933    0.052803   
1   0.109348  0.078518  0.141711     0.065447   0.017669    0.051654   
2   0.065568  0.046520  0.211722     0.059158   0.019231    0.023626   
3   0.226048  0.066617  0.042131     0.018345   0.003343    0.065762   
4   0.194714  0.067542  0.091198     0.023901   0.004486    0.054898   

   bc_course00  bc_course01  bc_course02  bc_course03  bc_course04  \
0     0.064837     0.095199     0.082009     0.040512     0.063252   
1     0.044893     0.071306     0.068422     0.032002     0.043180   
2     0.027656     0.045421     0.045238     0.025275     0.023810   
3     0.034319     0.043647     0.027090     0.054336     0.065374   
4     0.024390     0.034832     0.025124     0.039481     0.047557   

   bc_course05  bc_course06  bc_course07  bc_course08  bc_course09  \
0     0.053916     0.030534     0.046508     0.038328     0.163762   
1     0.036149     0.029388     0.042910     0.033625     0.152078   
2     0.027106     0.023077     0.032784     0.022711     0.202930   
3     0.045046     0.062653     0.097944     0.085507     0.098410   
4     0.033526     0.049596     0.080512     0.069337     0.085243   

   bc_course10  bc_course11  bc_course12  first  second  third  base_cnt  \
0     0.057728     0.130872     0.132543      0       0      0         0   
1     0.078157     0.203642     0.164248      0       0      0         0   
2     0.081319     0.285165     0.157509      0       0      0         0   
3     0.098527     0.055307     0.231840      0       0      0         0   
4     0.121054     0.076923     0.312424      0       0      0         0   

   top_bot  role  pos_pit  date_from_opening  ...  salary_year_bat  \
0        1     1        0                  0  ...           1875.0   
1        1     1        0                  0  ...           1875.0   
2        1     1        0                  0  ...           1875.0   
3        1     1        0                  0  ...            320.0   
4        1     1        0  

### train
- 行数: 257117

In [5]:
train = all_pitch.dropna(subset=['course'])
train.shape

(257117, 181)

### test
- 行数: 521650

In [6]:
test = all_pitch[all_pitch['course'].isnull()]
test.shape

(521650, 181)

In [7]:
del all_pitch
gc.collect()

60

In [8]:
train_d = train.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

In [9]:
test_d = test.drop([
    'No', 
    'course', 
    'ball'
], axis=1)

## Dataset作成

In [10]:
lgb_train = lgb.Dataset(train_d, train['course'])

### LGBM学習
- multiclass : クラス数=8
- tuning結果のパラメータを使用
- 'learning_rate' : 0.1

In [11]:
%%time
lgb_param = {
        'objective' : 'multiclass',
        'boosting_type': 'gbdt',
        'metric' : 'multi_logloss',
        'num_class' : 13,
        'seed' : 0,
        'learning_rate' : 0.1,
        'lambda_l1': 8.769293390201968, 
        'lambda_l2': 3.913949617576324e-05, 
        'num_leaves': 6, 
        'feature_fraction': 0.4, 
        'bagging_fraction': 0.8391111798378441, 
        'bagging_freq': 4, 
        'min_child_samples': 50
    }

# 学習
cv_results = lgb.cv(lgb_param, lgb_train,
                    num_boost_round=15000,
                    early_stopping_rounds=100,
                    verbose_eval=100,
                    nfold=4)

num_boost_round = len(cv_results['multi_logloss-mean'])
print('Best num_boost_round:', num_boost_round)
best_cv_score = cv_results['multi_logloss-mean'][-1]
print('Best CV score:', best_cv_score)

[100]	cv_agg's multi_logloss: 2.32538 + 0.000729413
[200]	cv_agg's multi_logloss: 2.32114 + 0.00078532
[300]	cv_agg's multi_logloss: 2.32039 + 0.000921575
[400]	cv_agg's multi_logloss: 2.32037 + 0.000983582
Best num_boost_round: 355
Best CV score: 2.3203553166835866
CPU times: user 40min 47s, sys: 8.3 s, total: 40min 55s
Wall time: 5min 8s


In [12]:
best_iter = int(num_boost_round * 1.1)

### 全体で再学習

In [13]:
%%time
lgb_model = lgb.train(lgb_param, lgb_train, num_boost_round=best_iter)

CPU times: user 11min 5s, sys: 2.07 s, total: 11min 7s
Wall time: 1min 23s


### Feature Importance

In [14]:
fi = lgb_model.feature_importance()
fn = lgb_model.feature_name()
df_feature_importance = pd.DataFrame({'feat_name':fn, 'feat_imp':fi})
df_feature_importance.sort_values('feat_imp', inplace=True)
df_feature_importance.tail(30)

feat_name  feat_imp
8    pitcher_cnt_in_inning       274
47             elapsed_min       276
22                 bc_fork       286
7      pitcher_cnt_in_game       292
169    salary_year_dif_b-c       300
103               course_5       306
102               course_4       323
115             height_bat       327
157            bmi_dif_p-b       330
101               course_3       332
99                course_1       337
106               course_8       342
105               course_7       342
30             bc_course04       350
31             bc_course05       353
107               course_9       353
98                course_0       354
29             bc_course03       355
110              course_12       358
109              course_11       365
108              course_10       369
173            bmi_dif_b-c       377
130                bmi_bat       377
132           bat_game_cnt       379
104               course_6       385
100               course_2       396
116             weight_bat       407
131             batter_cnt       457
46       date_from_opening       697
48                min_diff       805

In [15]:
%%time
predict = lgb_model.predict(test_d, num_iteration = lgb_model.best_iteration)

CPU times: user 2min 30s, sys: 756 ms, total: 2min 31s
Wall time: 20 s


## 結果出力

In [16]:
submit = pd.DataFrame(predict)
submit.reset_index(inplace=True)
print(submit.shape)
submit.head()

(521650, 14)


index         0         1         2         3         4         5  \
0      0  0.058623  0.131925  0.065878  0.037556  0.065146  0.030982   
1      1  0.053783  0.141413  0.068840  0.031136  0.060545  0.032488   
2      2  0.034906  0.086898  0.053487  0.024440  0.040331  0.023431   
3      3  0.044954  0.107299  0.062823  0.026353  0.043838  0.024614   
4      4  0.056951  0.130496  0.066417  0.035900  0.046994  0.030556   

          6         7         8         9        10        11        12  
0  0.055265  0.062547  0.039503  0.167777  0.090125  0.101743  0.092930  
1  0.048619  0.068480  0.035956  0.150735  0.081856  0.116585  0.109566  
2  0.044459  0.060658  0.028220  0.156206  0.105148  0.205024  0.136793  
3  0.044677  0.059516  0.035331  0.170849  0.108832  0.146827  0.124086  
4  0.048047  0.069077  0.044561  0.140288  0.080713  0.128541  0.121461

In [17]:
submit.mean()

index    260824.500000
0             0.046847
1             0.068777
2             0.056574
3             0.046458
4             0.059745
5             0.046970
6             0.043153
7             0.064068
8             0.052963
9             0.110210
10            0.088443
11            0.137125
12            0.178668
dtype: float64

In [18]:
submit.to_csv(SUBMIT, header=False, index=False)

### SIGNATE

In [19]:
'signate submit --competition-id=276 ./{} --note feat={}_cv={}'.format(SUBMIT, column_cnt, best_cv_score)

'signate submit --competition-id=276 ./submit/course_16.csv --note feat=181_cv=2.3203553166835866'

### GCS

In [20]:
'gsutil cp ./{} gs://yasunobu_baseball/submit/'.format(SUBMIT)

'gsutil cp ./submit/course_16.csv gs://yasunobu_baseball/submit/'

### 評価結果
#### 16(181)  ...選手IDごとの2017年のコース実績
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 355
- Best CV score: 2.3203553166835866
- Wall time: 5min 8s
- 評価結果  : 2.3608998

#### 15(153)  ...左右で球種/コースを別に集計
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 932
- Best CV score: 2.3387434142884445
- Wall time: 9min 41s
- 評価結果  : 2.3552543

#### 14(153)  ...左右で球種を別に集計
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 909
- Best CV score: 2.340082741044867
- Wall time: 9min 40s
- 評価結果  : 2.3555548

#### 3(157)  ...前の投球からの時間差を追加
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 857
- Best CV score: 2.340703597440808
- Wall time: 9min 20s
- 評価結果  : 2.3554087

#### 2(156)  ...merge12のパラメータをチューニング
- 'learning_rate' : 0.1,
- 'num_leaves': 6,
- Best num_boost_round: 835
- Best CV score: 2.3419206656914
- Wall time: 9min 7s
- 評価結果  : 2.3555299

#### 2(156)  ...merge12(2017のコースデータを付加) 
- 'learning_rate' : 0.1,
- Best num_boost_round: 121
- Best CV score: 2.344104226715559
- Wall time: 12min 30s
- 評価結果  : 2.3564326

#### 1(143)  ...球種予測の11と同じデータ 
- 'learning_rate' : 0.1,
- Best num_boost_round: 122
- Best CV score: 2.3445089434874253
- Wall time: 12min 2s (local)
- 評価結果  : 2.3562648